Add and refine guidelines

p4checo opened the pull request:
At long last, here's the update to our styleguide! 📝✅🚔

## Changes:

- Add new sections:
  + Documentation
  + Shadowing private members
  + Generics
  + Type aliases
  + Attributes
  + `switch` statements
- Add new guideline to use Swift 5.1 implicit returns on single line
closures.
- Refine existing guidelines:
  + `self` usage
  + `Void` function return
  + ternary operator.
  + `guard` statements.
- Make formatting improvements.
- Update code samples to 4 space indentation. 
- Merge some upstream changes
- Revise License and Credits

Feel free to comment/question anything! Ideally read the whole document for a clearer picture of the whole set of guidelines. 🙏

p4checo committed:
Merge upstream changes, add some adjustments

- Update Copyright notice in LICENSE.
- Update contributors and hat tip in CREDITS.
- Merge some upstream changes in README.

p4checo committed:
Add and refine guidelines

- Add new sections:

  + Documentation

  + Shadowing private members

  + Generics

  + Type aliases

  + Attributes

  + `switch` statements

- Add new guideline to use Swift 5.1 implicit returns on single line
closures.

- Refine existing guidelines:

  + `self` usage

  + `Void` function return

  + ternary operator.

  + `guard` statements.

- Make formatting improvements.

- Update code samples to 4 space indentation.

mickael-domingues reviewed (approved):
Good job 👏

portellaa commented on README.markdown:
```diff
@@ -365,34 +455,76 @@ The example above demonstrates the following style guidelines:
+let dynamicState: Property<Int>
+private let _dynamicState: MutableProperty<Int>
+
+func reticulateSplines(spline: [String])
```
Accordingly with `#Prose` section, shouldn't it be better to name this method `reticulate(splines: [String])`. The same applies to the private method bellow.

p4checo replied:
I think you mean more the `#Naming` section, perhaps?

Following Swift [API Design guidelines](https://swift.org/documentation/api-design-guidelines/) Argument Labels section below, I agree that we should change the signatures, but maybe we should omit the label since it's a part of a grammatical phrase:

![Screenshot 2019-10-24 at 12 16 49](https://user-images.githubusercontent.com/1391324/67480798-79a62f80-f658-11e9-98ec-639c0f6ba03c.png)

So, perhaps we should use `func reticulateSplines(_ splines: [String])` instead? 🤔

portellaa replied:
honestly, i think it's a decision for the team, i prefer `func reticulate(splines: ...` but `func reticulateSplines(_ splines: ...` it's more readable.

portellaa commented on README.markdown:
```diff
@@ -692,58 +979,62 @@ Code should not create reference cycles. Analyze your object graph and prevent s
+    guard let `self` = self else {
+        return
+    }
+    let model = strongSelf.updateModel(response)
```
typo: `strongSelf` instead of `self`

p4checo replied:
Well spotted! 🤓

portellaa reviewed (approved):
Nice additions 👏 🎉 

Just added minor comments and 💄 ones

filipe-lemos commented on README.markdown:
```diff
@@ -273,26 +306,28 @@ else {
+* Long lines should be wrapped at 120 characters.
 
-* Avoid trailing whitespaces at the ends of lines.
+* Remove any trailing whitespaces at the ends of lines.
```
Maybe add a **Pro Tip** about the Xcode feature to automatically trim these whitespaces?

p4checo replied:
We have a screenshot right next to this line, i.e. `screens/xcode_settings.png` that has the Xcode settings screen. 

However we can be more explicit and add the pro tip 🎩

filipe-lemos replied:
I was only looking at the diff and din't notice the screenshot - it's perfectly fine as it is 👌

filipe-lemos commented on README.markdown:
```diff
@@ -539,65 +813,79 @@ Static methods and type properties work similarly to global functions and global
+
+// another example
+UIView.animate(withDuration: 2.0) { [weak self] in
+    guard let strongSelf = self else { return }
```
Although it's under a **Not Preferred** example, should we follow `self = self` here?

p4checo replied:
I left it as was given it's **Not Preferred** like you say, but I guess we should focus on the topic at hand for each example, so I'll update to the correct `self` strongification 💪  idiom 😉

p4checo replied:
Actually, in this particular case we're discussing shadowing the original name whenever possible, so I think `strongSelf` is relevant for the *Not Preferred* example 🤔

filipe-lemos replied:
Oh yeah, it makes perfect sense, I didn't notice what it was examplifying. 😅
I guess `self = self` is a specific case of this more general shadowing rule.

p4checo replied:
precisely 😜

silvaric reviewed (approved):
Good one.
LGTM 👍

filipe-lemos commented on README.markdown:
```diff
@@ -756,147 +1047,306 @@ Only explicitly use `open`, `public`, and `internal` when you require a full acc
 
 Guard statements are required to exit in some way. Generally, this should be simple one line statement such as `return`, `throw`, `break`, `continue`, and `fatalError()`. Large code blocks should be avoided. If cleanup code is required for multiple exit points, consider using a `defer` block to avoid cleanup code duplication.
 
+if possible, place the exit closure inline with the `guard`, otherwise in a new line:
```
Uncapitalized _if_

filipe-lemos reviewed (approved):
Great refresh! 👏
Left some very minor comments. 🚀

p4checo committed:
Code review fixes + Improvements

- Code review changes

- Add Function spacing guideline.

- Add Raw strings guideline.

diogoeiras commented on README.markdown:
```diff
@@ -242,57 +254,86 @@ override func tableView(_ tableView: UITableView, numberOfRowsInSection section:
 
+## Spacing
+
+* Indent using **4 spaces** rather than tabs to conserve space and help prevent line wrapping. Check xcode settings at the end.
```
I'm a spaces guy but I do remember a discussion about accessibility of _spaces vs tabs_. More details can be found [here](https://www.reddit.com/r/javascript/comments/c8drjo/nobody_talks_about_the_real_reason_to_use_tabs/). What you think?

ricardomsm replied:
In our case, even with Xcode editing settings, does it still differ using spaces and tabs?

p4checo replied:
> In our case, even with Xcode editing settings, does it still differ using spaces and tabs?

yes, we can still configure Xcode to use tabs instead of spaces

I also prefer using spaces like we currently do, and given so far we haven't had anyone "complaining" that spaces aren't accessible for them, I would say we leave it as is until this becomes an actual problem. Also, this doesn't seem to have raised much discussion anyways 😆 

Nonetheless, these are just my 5¢ so feel free to continue the discussion.

ivopintodasilva commented on README.markdown:
```diff
@@ -365,34 +455,76 @@ The example above demonstrates the following style guidelines:
 
-Use self only when required by the compiler (in `@escaping` closures, or in initializers to disambiguate properties from arguments). In other words, if it compiles without `self` then omit it.
-
+Use `self` only when required by the compiler (e.g. in `@escaping` closures) or inside `init`'s when initializing each property (i.e. not just to disambiguate properties from arguments, but on **all** of them).
```
Maybe highlight the `weak self` unwrapping like `guard let self = self` as the preferred against `guard let strongSelf = self`?

ivopintodasilva replied:
It's further down the file, nevermind 🥳

p4checo replied:
**Pro Tip**: you can remove comments from a review before publishing it :trollface:

ivopintodasilva commented on README.markdown:
````diff
@@ -406,54 +538,190 @@ In the below example, `Box` has a particular purpose and customization in a deri
 ```swift
 func reticulateSplines(spline: [Double]) -> Bool {
-  // reticulate code goes here
+    // reticulate code goes here
````
Missing a new line between the code and function declaration 💄

p4checo replied:
Well spotted! Fixed

p4checo committed:
More code review fixes
